Update the docs of CKB Bitcoin SPV Type Script (#28)
ckb-cell/ckb-bitcoin-spv-contracts: contracts/ckb-bitcoin-spv-type-lock/README.md

> # CKB Bitcoin SPV Type Lock

Reviewer: ```suggestion
# CKB Bitcoin SPV Type Script
```

Suggest changing the filename as well

A type script designed for Bitcoin SPV clients ensures the valid synchronization of the [Bitcoin] blockchain state into the Nervos [CKB] network.

## Brief Introduction

> A Bitcoin  SPV on CKB consists of cells that are managed by the CKB Bitcoin SPV Type Lock and identified by the script args. The number of live cells with the script args remains fixed once created, and these cells will be destroyed collectively as a group.

Reviewer 2: ```suggestion
A Bitcoin  SPV on CKB consists of cells that are managed by the CKB Bitcoin SPV Type Script and identified by the script args. The number of live cells with the script args remains fixed once created, and these cells will be destroyed collectively as a group.
```

      - target adjust info
  ```

- **SPV Info Cell**

>   The SPV info cell stores the basic information of the current Bitcoin SPV instance, such as `tip_ client_cell_id`. Each Bitcoin SPV

Reviewer 2: ```suggestion
  The SPV info cell stores the basic information of the current Bitcoin SPV instance, such as `tip_client_id`. Each Bitcoin SPV
```

### Operations

There are 4 kinds of operations in the Bitcoin SPV type script:

- **Create**

  This operation initiates all necessary cells for a Bitcoin SPV instance in a single transaction.

>   The `outputs` include one SPV info cell and at least one SPV client cell. Cells should be consecutive, with the SPV info cell first,

Reviewer 2: ```suggestion
  The `outputs` include one SPV info cell and at least three SPV client cell. Cells should be consecutive, with the SPV info cell first,
```

See https://github.com/KryStar11/ckb-bitcoin-spv-contracts/blob/770049914b6818f51b53c11be434a43876b87c45/contracts/ckb-bitcoin-spv-type-lock/src/operations/create.rs#L13-L39

### Usages

To verify a transaction using the Bitcoin SPV Client cell, follow these steps:

- Select an SPV client cell that contains the block where the transaction is;

- Create a transaction proof, with the following data:

  - The MMR proof of the block header which contains this transaction;

  - The TxOut proof of the transaction;

  - The index of the transaction;

  - The height of the block header.

- Use the SpvClient::verify_transaction(..) for the verification. For detailed guidance, please refer to the [API example].

### Limits

- The minimum count of SPV client cells is 3;
  
- While there is no fixed maximum count of SPV client cells; it is advisable not to exceed `250` given the **`u8`** data type.

### Known Issues and Solutions

- **Issue #2**: `VM Internal Error: MemWriteOnExecutablePage`

  **Solution**: Don't set hash type[^1] to be `Data`.

>   For the appropriate use of hash type, refer to the "Code Locating" section in [CKB RFC 0022]. `Data1` is
>   introduced in [CKB RFC 0032], and `Data2` is introduced in [CKB RFC 0051].

Reviewer: ```suggestion
  `Data1` is introduced in [CKB RFC 0032], and `Data2` is introduced in [CKB RFC 0051].
```

Duplicates the link:

> Don't set hash type[^1]

- **Issue #2**: Failed to reorg when there is only 1 stale SPV client.

  **Solution**: When only one SPV client cell is stale, a typical reorg transaction has the same structure as an update transaction,

>   consisting of one SPV client cell in the inputs and one SPV client cell in the outputs. This similarity can lead to ambiguity.

Reviewer: ```suggestion
  consisting of one SPV client cell in the inputs and one SPV client cell in the outputs. However, this similarity can lead to ambiguity.
```

  To address this issue, the following rules have been set: 

>     - In cases where only one SPV client has failed, the reorg transaction must involve the reconstruction of one separate SPV clients;

Reviewer: ```suggestion
    - In cases where only one SPV client has failed, the reorg transaction must involve the reconstruction of one additional  SPV client;
```

>     - Specifically, the reorg transaction for one stale SPV client should include two SPV client cell in the `inputs` and two SPV client

Reviewer: ```suggestion
    Specifically, the reorg transaction for one stale SPV client should include two SPV client cell in the `inputs` and two SPV client
```

    cell in the `outputs` ;

>     - Considering that reorgs are a rare occurrence on the Bitcoin mainnet, the cost incurred by this approach is considered manageable.

Reviewer: ```suggestion
    Considering that reorgs are a rare occurrence on the Bitcoin mainnet, the cost incurred by this approach is considered manageable.
```

### Limits

- The minimum count of SPV client cells is 3;
  
- While there is no fixed maximum count of SPV client cells; it is advisable not to exceed `250` given the **`u8`** data type.

### Known Issues and Solutions

- **Issue #1**: `VM Internal Error: MemWriteOnExecutablePage`

  **Solution**: Don't set hash type[^1] to be `Data`.

  `Data1` is introduced in [CKB RFC 0032], and `Data2` is introduced in [CKB RFC 0051].

- **Issue #2**: Failed to reorg when there is only 1 stale SPV client.

  When only one SPV client cell is stale, a typical reorg transaction has the same structure as an update transaction,
  consisting of one SPV client cell in the inputs and one SPV client cell in the outputs. However, this similarity can lead to ambiguity.

  **Solution**: To address this issue, the following rule has been set: 

    - In cases where only one SPV client has failed, the reorg transaction must involve the reconstruction of one additional SPV client;
    Specifically, the reorg transaction for one stale SPV client should include two SPV client cell in the `inputs` and two SPV client

>     cell in the `outputs` ;

Reviewer 2: ```suggestion
    cells in the `outputs` ;
```
